golang/cosmos/x/vstorage/vstorage_test.go

```diff
@@ -0,0 +1,258 @@
+package vstorage
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+
+	"github.com/cosmos/cosmos-sdk/store"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	agorictypes "github.com/Agoric/agoric-sdk/golang/cosmos/types"
+	"github.com/Agoric/agoric-sdk/golang/cosmos/vm"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var (
+	storeKey = storetypes.NewKVStoreKey(types.StoreKey)
+)
+
+func ptr[T any](v T) *T {
+	return &v
+}
+
+type testKit struct {
+	keeper  Keeper
+	handler vstorageHandler
+	ctx     sdk.Context
+	cctx    *vm.ControllerContext
+}
+
+func makeTestKit() testKit {
+	keeper := NewKeeper(storeKey)
+	db := dbm.NewMemDB()
+	ms := store.NewCommitMultiStore(db)
+	ms.MountStoreWithDB(storeKey, sdk.StoreTypeIAVL, db)
+	err := ms.LoadLatestVersion()
+	if err != nil {
+		panic(err)
+	}
+	ctx := sdk.NewContext(ms, tmproto.Header{}, false, log.NewNopLogger())
+	cctx := &vm.ControllerContext{Context: ctx}
+	handler := vstorageHandler{keeper}
+	return testKit{keeper, handler, ctx, cctx}
+}
+
+func callReceive(
+	handler vstorageHandler,
+	cctx *vm.ControllerContext,
+	method string,
+	args []interface{},
+) (string, error) {
+	var rawArgs []json.RawMessage
+	for _, arg := range args {
+		rawArg, _ := json.Marshal(arg)
+		rawArgs = append(rawArgs, json.RawMessage(rawArg))
+	}
+	req := vstorageMessage{method, rawArgs}
+	reqBytes, _ := json.Marshal(req)
+	return handler.Receive(cctx, string(reqBytes))
+}
+
+func TestGetAndHas(t *testing.T) {
+	kit := makeTestKit()
+	keeper, handler, ctx, cctx := kit.keeper, kit.handler, kit.ctx, kit.cctx
+
+	keeper.SetStorage(ctx, types.NewStorageEntry("foo", "bar"))
+	keeper.SetStorage(ctx, types.NewStorageEntry("empty", ""))
+	keeper.SetStorage(ctx, types.NewStorageEntryWithNoData("special.nodata"))
```

Comment:
nit: I'd prefer if the test was more realistic and no data entries only existed for cases where there are nested entries

```suggestion
	keeper.SetStorage(ctx, types.NewStorageEntry("placeholder.empty", ""))
	keeper.SetStorage(ctx, types.NewStorageEntryWithNoData("placeholder"))
```

Reply:
Fixed (in both places).